src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                var dummyImport = CreateImport(generator, "Dummy", withFormatterAnnotation: false);
+
+                foreach (var refLoc in refLocations)
+                {
+                    Debug.Assert(document.Id == refLoc.Document.Id);
+
+                    // Ignore references via alias. For simple cases where the alias is defined as the type we are interested,
+                    // it will be handled properly because it is one of the reference to the type symbol. Otherwise, we don't
+                    // attempt to make a potential fix, and user might end up with errors as a result.                    
+                    if (refLoc.Alias != null)
+                    {
+                        continue;
+                    }
+
+                    // Other documents in the solution might have changed after we calculated those ReferenceLocation, 
+                    // so we can't trust anything to be still up-to-date except their spans.
+
+                    // Get inner most node in case of type used as a base type. e.g.
+                    //
+                    //      public class Foo {}
+                    //      public class Bar : Foo {}
+                    //
+                    // For the reference to Foo where it is used as a base class, the BaseTypeSyntax and the TypeSyntax
+                    // have exact same span.
+
+                    var refNode = root.FindNode(refLoc.Location.SourceSpan, getInnermostNodeForTie: true);
+                    if (syncNamespaceService.TryGetReplacementReferenceSyntax(
+                        refNode, namespaceParts, syntaxFacts, out var oldNode, out var newNode))
+                    {
+                        editor.ReplaceNode(oldNode, newNode.WithAdditionalAnnotations(Simplifier.Annotation));
+                    }
+
+                    // Use a dummy import node to figure out which container the new import will be added to.
+                    var container = addImportService.GetImportContainer(root, refNode, dummyImport);
+                    containers.Add(container);
+                }
+
+                foreach(var container in containers)
+                {
+                    editor.TrackNode(container);
+                }
+
+                document = editor.GetChangedDocument();
+                root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                return (document, containers.Select(c => root.GetCurrentNode(c)).ToImmutableArray());
+            }
+
+            private async Task<Solution> RemoveUnnecessaryImportsAsync(
+                Solution solution, 
+                ImmutableArray<DocumentId> ids, 
+                ImmutableArray<string> names, 
+                CancellationToken cancellationToken)
+            {
+                var LinkedDocumentsToSkip = PooledHashSet<DocumentId>.GetInstance();
+                foreach (var id in ids)
```

Comment:
this work seems like it can be done in parallel.

Reply:
Once I get all changed documents in parallel, how do I merge them into one solution?

Reply:
I guess something like this?
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs,52

Reply:
Exactly.  The expensive part is getting the changed documentd.  Iterating over them, updating the solution with each document is then fast (it basically just forks the solution with that single document replaced in it).